### PR TITLE
Provide global configuration mechanism

### PR DESCRIPTION
**Pull request by lorenzocelli**
Issue #34

This solution uses a `Config` class defining as attributes the tool2schema-related settings. Currently, the available settings are `tags` and `ignored_parameters`.

```python
class Config:
    """
    Configuration class for tool2schema.
    """

    def __init__(self, parent: Optional[Config] = None, **settings):
        self.parent = parent
        self.settings = settings
        self.tags: list[str] = self._get_setting("tags", [])
        self.ignored_parameters: list[str] = self._get_setting(
            "ignored_parameters", ["self", "args", "kwargs"]
        )

    def _get_setting(self, name: str, default):
        if not self.parent:
            return self.settings.pop(name, default)

        return self.settings.pop(name, getattr(self.parent, name))
```

By creating an instance of `Config` without any additional parameters one gets the default configuration. 
The default values are specified in the constructor and passed to `_get_setting`. 

The `Config` class can accept a parent configuration which is used (if present) to retrieve the settings that are not overridden by keyword arguments. 

In the following example, `my_config` will have the same settings as the parent `parent_config`, except for `tags` which will be overridden.

```python
my_config = Config(parent_config, tags=["hello"])
```

In  `__init__.py` an instance of `Config` is created to serve as the global configuration:
```python
# Default global configuration
CONFIG = Config()
```

This allows the user to edit the settings as follows:
```python
import tool2schema

tool2schema.CONFIG.ignored_parameters = ["b", "c"]
```

### CODE REVIEW

**Comment by LorenzoCelli** [comment]
Marked this as a draft because I am not sure about two things.

1. I don't know if we want to keep `tags` as part of the configuration. The only use case I see for a globally-set tag is if you wanted to tag every method except a few.
2. Not sure if we should go with names such as `ignored_parameters` which sound more like settings but sound worse as parameters of `GPTEnabled`, or use the imperative `ignore_parameters`
```python
@GPTEnabled(ignored_parameters=["a", "d"])
def function():
    pass

@GPTEnabled(ignore_parameters=["a", "d"])
def function():
    pass
```

@siliconlad let me know what you think 😄

**Comment by siliconlad** [comment]
1. Yeah I don't think having a global tags configuration makes much sense. I think there are a potentially a case to be made for filters of specific tags. We can do that in a different issue when the need for it arises.
2. Could we have `ignored_parameters` for the configuration, but `ignore_parameters` for `GPTEnabled`? Is that too confusing?

**Comment by LorenzoCelli** [comment]
> 1. Yeah I don't think having a global tags configuration makes much sense. I think there are a potentially a case to be made for filters of specific tags. We can do that in a different issue when the need for it arises.
> 2. Could we have `ignored_parameters` for the configuration, but `ignore_parameters` for `GPTEnabled`? Is that too confusing?

Sure, I'll remove the tags from the configuration then.

As for 2, yes we can do it with a little more code, something like:

```python
self.config = Config(tool2schema.CONFIG,
    ignored_parameters=kwargs.pop("ignore_parameters", None),
    **kwargs
)
```

But yes I guess it could be a little confusing for the user, also considering that the names only differ by one letter.

I would say we stick to the imperative style which is presumably a more common use case, and if we find it necessary we can always change the names in the configuration later. What do you think?

**Comment by siliconlad** [comment]
Sounds good to me!

**Comment by LorenzoCelli** [comment]
I am keeping this branch up to date with `feature/support-enumeration-type-hints` so that I can store the configuration as a class attribute in `FunctionSchema`. It's preferable to wait until `feature/support-enumeration-type-hints` is merged into main before merging this branch.

File: tool2schema/__init__.py
Comment (siliconlad) [inline tool2schema/__init__.py:24]: If this method is included here then, it shouldn't have the `_` prefix. But perhaps a better place would be within the `Config` object to facilitate an API like

```python
tool2schema.CONFIG.reset_default()
```

**Review by siliconlad — CHANGES_REQUESTED** [review]
Overall, looks excellent!

I would also add something to the README.

File: tool2schema/__init__.py
Comment (lorenzocelli) [reply tool2schema/__init__.py:24]: I added the prefix `_` because it was meant to be an internal method used only for testing purposes. With the current configuration class it is not possible to reset the configuration because the attributes were meant to be immutable and the default values are only know in the constructor.

But here's another idea, using properties instead of class attributes:
```python
class Config:

    def __init__(self, parent: Optional[Config] = None, **settings):
        self._parent = parent
        self._settings = settings

    @property
    def ignore_parameters(self) -> list[str]:
        return self._get_setting(Config.ignore_parameters.fget.__name__, ["self", "args", "kwargs"])

    @ignore_parameters.setter
    def ignore_parameters(self, value: list[str]):
        self._set_setting(Config.ignore_parameters.fget.__name__, value)

    def reset(self):
        self._settings = {}

    def _get_setting(self, name: str, default):
        default = default if not self._parent else getattr(self._parent, name)
        return self._settings.get(name, default)

    def _set_setting(self, name: str, value):
        self._settings[name] = value
```

Advantages:
- Settings are dynamic, so if you change the parent configuration it will reflect on children
- We can document the properties
- We can have a reset method
- There are no strings with the name of the settings, thus the properties can be renamed easily

Disadvantage:
- You have to write 7 lines of code and two methods (getter/setter) to add a new setting.

Comment (lorenzocelli) [reply tool2schema/__init__.py:24]: Let me know! :)

Comment (lorenzocelli) [reply tool2schema/__init__.py]: I checked it in because I liked it better but I can always go back

Comment (siliconlad) [reply tool2schema/__init__.py]: Yeah, I like it. To me the main trade-off is

We have autocomplete for the `Config` object
vs
Two methods per setting

I think that's acceptable.

**Comment by lorenzocelli** [comment]
I have updated the README with some details on how to change the global configuration

File: tool2schema/config.py
Comment (siliconlad) [inline tool2schema/config.py:30]: Should we reset to `{}` or to the `settings` argument in the constructor?

File: tool2schema/config.py
Comment (siliconlad) [inline tool2schema/config.py:42]: Should we worry if `name` isn't in `self._parent`? Such that we can return `default`?

**Review by siliconlad — CHANGES_REQUESTED** [review]
Couple of comments :)

File: tool2schema/config.py
Comment (lorenzocelli) [reply tool2schema/config.py:42]: I don't think so, `getattr` should get the property from the parent which will in turn return either its setting or the default

File: tool2schema/config.py
Comment (lorenzocelli) [reply tool2schema/config.py:30]: It depends, do we want it to reset everything or reset it to the settings given in the constructor?
```python
config = Config(ignore_parameters=True)
config.ignore_function_description=True
config.reset_default() # should this revert only 'ignore_function_description' or also 'ignore_parameters'?
```
In the case of the global configuration it makes no difference

File: tool2schema/__init__.py
Comment (lorenzocelli) [reply tool2schema/__init__.py]: Done in a05e765

File: tool2schema/config.py
Comment (siliconlad) [reply tool2schema/config.py:30]: To me, only reverting `ignore_function_description` makes sense to me.

File: tool2schema/config.py
Comment (siliconlad) [reply tool2schema/config.py:42]: Should this then not be `getattr(self._parent, name, default)`?

File: tool2schema/config.py
Comment (lorenzocelli) [reply tool2schema/config.py]: Done with 5a6b214, added two tests to check this

File: tool2schema/config.py
Comment (lorenzocelli) [reply tool2schema/config.py:44]: `getattr` cannot fail to return a value, when the parent is present it will either return the default or ask its parent which will do the same (some parent will return a value eventually)

Comment (siliconlad) [reply tool2schema/config.py:44]: How does `getattr` know its parent when it is a built-in Python function?

Comment (lorenzocelli) [reply tool2schema/config.py:44]: It doesn't, I'm passing the parent as an argument

Comment (lorenzocelli) [reply tool2schema/config.py:44]: If the parent is not present getattr will not be invoked

Comment (siliconlad) [reply tool2schema/config.py:44]: Ah yes because `getattr` will call the `@property` method.

**Review by siliconlad — APPROVED** [review]
LGTM!

**Pull request merged**
